### ISSUE TITLE
Malformed credentials file

### ISSUE THREAD

**Issue by migueloller** [pulumi/pulumi#3877]
Running Pulumi CLI commands in CI is failing with the following error:

```
error: could not get cloud url: unmarshalling credentials file: unexpected end of JSON input
```

The following scripts are representative of what was running when the error happened:

```bash
# prepare.sh

curl https://sdk.cloud.google.com | bash > /dev/null
export PATH=$PATH:/root/google-cloud-sdk/bin
KEY_FILE="$(mktemp)"
echo "$GCP_SERVICE_ACCOUNT" > "$KEY_FILE"
gcloud auth activate-service-account --key-file "$KEY_FILE"

curl -fsSL https://get.pulumi.com/ | bash
export PATH=$PATH:$HOME/.pulumi/bin
pulumi login
```

```bash
# deploy.sh

pulumi stack init "$PULUMI_STACK" || true

if [[ "$PULUMI_STACK" != "production" ]] || [[ "$PULUMI_STACK" != "staging" ]]; then
  pulumi config refresh --stack development || true
  json="$(pulumi config --stack development --show-secrets --json)"
  keys="$(echo $json | jq -r 'keys[]')"

  for key in $(echo $keys | tr '\n' ' ')
  do
    value="$(echo $json | jq -r '.["'"$key"'"].value')"
    is_secret="$(echo $json | jq -r '.["'"$key"'"].secret')"
    pulumi config --stack "$PULUMI_STACK" set "$key" "$value" $(if [ $is_secret = 'true' ]; then echo '--secret'; fi)
  done
fi
```

Some things to keep in mind:
- In some instances, the error doesn't happen at all.
- The `prepare.sh` bash script is run once before updating each stack.
- The `deploy.sh` bash script above executes in parallel for multiple Pulumi projects. In our case there are around 9 projects.
- The scripts were modified in some runs to output the Google key file and well as the Pulumi credentials file (`$HOME/.pulumi/credentials.json`) before running the scripts and they seemed to always be valid JSON.
- The scripts were modified to trace each command and the error seemed to happen during the `pulumi config set` call in `deploy.sh`.
- There is a conversation in the Pulumi Slack in the channel `#gcp` that precludes this issue.

Evaluating all the information throughout the conversation and presented above a potential cause for the issue could be concurrent use of Pulumi commands in the CLI. If the commands are writing to the credentials file while others read it, it's possible that if the writes are not atomic, the reads would be in a transient state that is malformed.

**Comment by migueloller** [comment]
I will be modifying the CI configuration so that it runs all projects in series and report back if that reliably stops the error from happening.

**Comment by migueloller** [comment]
Reporting back. It seems that running the projects serially fixed the issue. This means that the likelihood of Pulumi commands not being safe to run in parallel is pretty high.

cc: @lukehoban

**Comment by alikhajeh1** [comment]
I ran into this issue whilst running parallel pulumi CLI commands too. The error is raised by https://github.com/pulumi/pulumi/blob/master/sdk/go/common/workspace/creds.go#L173. Following this function shows that the pulumi CLI's login method writes the creds to the `credentials.json` file from the env vars. That was a little bit surprising but it serves a purpose; for example, it keeps track of `lastValidatedAt` in the `credentials.json` file. I suspect that parallel writes are causing some of the file reads to be corrupted due to partial writes.

I found a simple workaround that others might find helpful: https://github.com/pulumi/pulumi/blob/master/sdk/go/common/workspace/creds.go#L29-L32
So the env `PULUMI_CREDENTIALS_PATH` can be used to separate each stack's `credentials.json` path using something like `/tmp/${uuid}`.
